tests/unit/test_user_traits.cc

```diff
@@ -0,0 +1,204 @@
+    EXPECT_FALSE((S::template has_traits<ShallowTrait>::value));
+
+    EXPECT_EQ(
+      (S::template has_traits<TraitPairA>::value),
+      (S::template has_traits<TraitPairB>::value)
+    );
+
+    if constexpr(S::template has_traits<CheckpointTrait>::value){
+      s | name;
+    }
+    s | val_a;
+  }
+
+  char name;
+  int val_a;
+};
+
+template<
+  typename S,
+  typename = typename S::template has_traits<CheckpointTraitNonintrusive>::type
+>
+void serialize(S& s, UserObjectA& obj){
+  s | obj.name;
+  obj.serialize(s);
+}
+
+namespace CheckpointNamespace {
+template<
+  typename S,
+  typename = typename S::template has_traits<CheckpointTraitNamespaced>::type
+>
+void serialize(S& s, UserObjectA& obj){
+  s | obj.name;
+  obj.serialize(s);
+}
+} // CheckpointNamespace
+
+/*
+ * Test that trait handling works from different sources
+ */
+TYPED_TEST_P(TestUserTraits, test_serialize_extra) {
+  using Trait = TypeParam;
+
+  UserObjectA objA(u_val);
+  char new_name = 'Z', old_name = objA.name;
+  ASSERT_NE(new_name, old_name);
+
+  //Without the checkpoint trait, name shouldn't be checkpointed.
+  auto serA = checkpoint::serialize(objA);
+  objA.name = new_name;
+  checkpoint::deserializeInPlace(serA->getBuffer(), &objA);
+  EXPECT_EQ(objA.name, new_name);
+  EXPECT_NE(objA.name, old_name);
+
+  //With the checkpoint trait, name change should be undone
+  objA.name = old_name;
+  serA = checkpoint::serialize<UserObjectA, Trait>(objA);
+  objA.name = new_name;
+  checkpoint::deserializeInPlace<UserObjectA, Trait>(serA->getBuffer(), &objA);
+  EXPECT_EQ(objA.name, old_name);
+  EXPECT_NE(objA.name, new_name);
+}
+REGISTER_TYPED_TEST_CASE_P(TestUserTraits, test_serialize_extra);
+using TraitTypes = ::testing::Types<CheckpointTrait, CheckpointTraitNonintrusive, CheckpointNamespace::CheckpointTraitNamespaced>;
+INSTANTIATE_TYPED_TEST_CASE_P(checkpoint_trait, TestUserTraits, TraitTypes, );
+
+
+
+struct UserObjectB : public UserObjectA {
+  UserObjectB() = default;
+  explicit UserObjectB(int val)
+    : UserObjectA(0, 'B'), val_b(val) {};
+
+  template <typename S>
+  void serialize(S& s) {
+    std::cout << "B: serializing with type "
+      << abi::__cxa_demangle(typeid(s).name(), nullptr, nullptr, nullptr)
+      << std::endl;
+    auto new_s = s.template withoutTraits<ShallowTrait>();
+    if (S::template has_traits<TraitPairA>::value){
+      auto newer_s = new_s.template withTraits<TraitPairB>();
+      UserObjectA::serialize(newer_s);
+    } else {
+      UserObjectA::serialize(new_s);
+    }
+
+    s | val_b;
+  }
+
+  int val_b;
+};
+
+TEST(TestUserTraits, test_trait_removal) {
+  UserObjectB objB(u_val);
+  int old_b_val = objB.val_b, old_a_val = objB.val_a;
+  int new_b_val = old_b_val+1, new_a_val = old_a_val+1;
+
+  auto serB = checkpoint::serialize<UserObjectB, ShallowTrait>(objB);
+  objB.val_b = new_b_val;
+  objB.val_a = new_a_val;
+  checkpoint::deserializeInPlace<UserObjectB, ShallowTrait>(serB->getBuffer(), &objB);
```

Comment:
should this test also check name to ensure that the UserObjectA was called correctly

Reply:
We can, but I don't think that's checking for anything that wouldn't already be caught by test_serialize_extra so I'd rather keep the checks within the test's scope

Reply:
With the extra check for correct trait detection, we should be fine leaving the name checking to the test_serialize_extra test